scipy/sparse/tests/test_base.py

```diff
@@ -5291,32 +5337,31 @@ def test_downcast_intp(self):
         # due to use of functions that only work with intp-size
         # indices.
 
-        @with_64bit_maxval_limit(fixed_dtype=np.int64,
-                                 downcast_maxval=1)
+        @with_64bit_maxval_limit(fixed_dtype=np.int64, downcast_maxval=1)
         def check_limited():
             # These involve indices larger than `downcast_maxval`
             a = csc_matrix([[1, 2], [3, 4], [5, 6]])
-            assert_raises(AssertionError, a.getnnz, axis=1)
+            assert_raises(AssertionError, a.count_nonzero, axis=1)
             assert_raises(AssertionError, a.sum, axis=0)
 
             a = csr_matrix([[1, 2, 3], [3, 4, 6]])
-            assert_raises(AssertionError, a.getnnz, axis=0)
+            assert_raises(AssertionError, a.count_nonzero, axis=0)
 
             a = coo_matrix([[1, 2, 3], [3, 4, 5]])
-            assert_raises(AssertionError, a.getnnz, axis=0)
+            assert_raises(AssertionError, a.count_nonzero, axis=0)
```

Comment:
To keep the same coverage, we should have:

```
CSR.sum(axis=1)
CSC.sum(axis=0)
COO.sum_duplicates()
```

Reply:
*Context for those not at the sparse array meeting:
A few lines above this suggestion is a comment stating that the purpose of these tests is to check that the np.bincount and np.reduceat code deals with intp downcasts so those functions work on 32-bit systems when using 64-bit indices.*

It looks like testing `COO.sum_duplicates()` actually **expands** the coverage. We didn't used to test the `reduceat` that is used in COO.sum_duplicates. And including it created errors. The errors are fixed by using `downcast_intp_index` in COO.sum_duplicates. So I have added those tests and fixed the errors that it found! Thanks for this pointer!

Secondly, I think we weren't seeing the main branch when we were searching for `bincount` at the meeting today. On the main branch, the function `count_nonzero` uses bincount (since June 4 -- PR 20857 and it will be released in v1.15). The new `axis` parameter in `count_nonzero` mimics the code in `_getnnz` quite closely. Replacing the `getnnz` test with `count_nonzero` test is thus both needed and desired since the code is the same in terms of testing index downcasting.

So, my commit to address this suggestion: 
- **bincount**: keeps the `count_nonzero` tests for CSC, CSR and COO as added in this PR
- **reduceat**: keeps the tests of `CSC.sum(axis=0)`, adds the test of `CSR.sum(axis=1)` (probably left out because it is the same code).
- **reduceat**: adds the `COO.sum_duplicates` tests for COO. To make this pass, we must fix the downcasting problem in `COO.sum_duplicates` by adding `downcast_intp_index`.

I also noticed that the `sum_duplicates` test will require that `has_canonical_format` be set to False to make the test actually call `reduceat` so I added that.

In summary we will test:
```python
CSR.sum(axis=1)
CSR.count_nonzero(axis=0)
CSC.sum(axis=0)
CSC.count_nonzero(axis=1)
has_canonical_format = False
COO.sum_duplicates()
COO.count_nonzero(axis=0)   # no need to test both axis=0 and axis=1 because same situation in both
```